// nmr-predictor/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.twoD = exports.spinus = exports.carbon = exports.proton = exports.fetchCarbon = exports.fetchProton = undefined;

var _superagent = require('superagent');

var _superagent2 = _interopRequireDefault(_superagent);

var _normalizeOptions = require('./normalizeOptions');

var _normalizeOptions2 = _interopRequireDefault(_normalizeOptions);

var _queryByHose = require('./queryByHose');

var _queryByHose2 = _interopRequireDefault(_queryByHose);

var _spinus = require('./spinus');

var _spinus2 = _interopRequireDefault(_spinus);

var _twoD = require('./twoD');

var _twoD2 = _interopRequireDefault(_twoD);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const defaultProtonUrl = 'https://raw.githubusercontent.com/cheminfo-js/spectra/master/packages/nmr-predictor/data/h1.json';
const defaultCarbonUrl = 'https://raw.githubusercontent.com/cheminfo-js/spectra/master/packages/nmr-predictor/data/nmrshiftdb2-13c.json';

const databases = {};

function fetchProton(url = defaultProtonUrl, dbName = 'proton') {
    return fetch(url, dbName, 'proton');
}

function fetchCarbon(url = defaultCarbonUrl, dbName = 'carbon') {
    return fetch(url, dbName, 'carbon');
}

function fetch(url, dbName, type) {
    if (databases[dbName] && databases[dbName].type === type && databases[dbName].url === url) {
        if (databases[dbName].fetching) {
            return databases[dbName].fetching;
        }
        return Promise.resolve(databases[dbName].db);
    }
    const database = {
        type,
        url,
        db: null,
        fetching: null
    };
    databases[dbName] = database;
    const fetching = _superagent2.default.get(url).then(res => {
        const db = res.body ? res.body : JSON.parse(res.text);
        database.db = db;
        database.fetching = false;
        return db;
    }).catch(e => {
        delete databases[dbName];
        throw e;
    });
    database.fetching = fetching;
    return fetching;
}

function proton(molecule, options) {
    options.atomLabel = 'H';
    [molecule, options] = (0, _normalizeOptions2.default)(molecule, options);
    const db = getDb(options.db || 'proton', 'proton');
    return (0, _queryByHose2.default)(molecule, db, options);
}

function carbon(molecule, options) {
    options.atomLabel = 'C';
    [molecule, options] = (0, _normalizeOptions2.default)(molecule, options);
    const db = getDb(options.db || 'carbon', 'carbon');
    return (0, _queryByHose2.default)(molecule, db, options);
}

function getDb(option, type) {
    if (typeof option === 'object') return option;
    if (typeof option !== 'string') throw new TypeError('database option must be a string or array');
    const db = databases[option];
    if (!db) throw new Error(`database ${option} does not exist. Did you forget to fetch it?`);
    if (db.fetching) throw new Error(`database ${option} is not fetched yet`);
    if (db.type !== type) throw new Error(`database ${option} is of type ${db.type} instead of ${type}`);
    return db.db;
}

exports.fetchProton = fetchProton;
exports.fetchCarbon = fetchCarbon;
exports.proton = proton;
exports.carbon = carbon;
exports.spinus = _spinus2.default;
exports.twoD = _twoD2.default;
